README.md
# NAME

Filesys::Notify::Light - Light and simple file system watcher

# SYNOPSIS

    use Filesys::Notify::Light;

    my $watcher = Filesys::Notify::Light->new([ "." ]);
    $watcher->wait(sub {
        for my $event (@_) {
            $event->{path} # full path of the file updated
        }
    });

# DESCRIPTION

Filesys::Notify::Light is a Light but unified interface to get
notifications of changes to a given filesystem path. It utilizes
inotify2 on Linux, fsevents on OS X, kqueue on FreeBSD and
FindFirstChangeNotification on Windows if they're installed, with a
fallback to the full directory scan if they're not available.

There are some limitations in this module. If you don't like it, use
[File::ChangeNotify](http://search.cpan.org/perldoc?File::ChangeNotify).

- There is no file name based filter. Do it in your own code.
- You can not get types of events (created, updated, deleted).
- Currently `wait` method blocks.

In return, this module doesn't depend on any non-core
modules. Platform specific optimizations with [Linux::Inotify2](http://search.cpan.org/perldoc?Linux::Inotify2),
[Mac::FSEvents](http://search.cpan.org/perldoc?Mac::FSEvents), [Filesys::Notify::KQueue](http://search.cpan.org/perldoc?Filesys::Notify::KQueue) and [Win32::ChangeNotify](http://search.cpan.org/perldoc?Win32::ChangeNotify)
are truely optional.

NOTE: Using [Win32::ChangeNotify](http://search.cpan.org/perldoc?Win32::ChangeNotify) may put additional limitations.

- [Win32::ChangeNotify](http://search.cpan.org/perldoc?Win32::ChangeNotify) uses FindFirstChangeNotificationA so that
Unicode characters can not be handled.
On cygwin (1.7 or later), Unicode characters should be able to be handled
when [Win32::ChangeNotify](http://search.cpan.org/perldoc?Win32::ChangeNotify) is not used.
- If more than 64 directories are included under the specified paths,
an error occurrs.

# AUTHOR

Tatsuhiko Miyagawa <miyagawa@bulknews.net>

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[File::ChangeNotify](http://search.cpan.org/perldoc?File::ChangeNotify) [Mac::FSEvents](http://search.cpan.org/perldoc?Mac::FSEvents) [Linux::Inotify2](http://search.cpan.org/perldoc?Linux::Inotify2) [Filesys::Notify::KQueue](http://search.cpan.org/perldoc?Filesys::Notify::KQueue)
[Win32::ChangeNotify](http://search.cpan.org/perldoc?Win32::ChangeNotify)


Changes
Filesys::Notify::Light (0.2.00)

  * Initial release from rewriting Filesys::Notify::Simple

-- Marcel Greter <perl-file-notify@ocbnet.ch>  Wed Mar 20 22:53:18 CET 2019


Makefile.PL
# Copyright (c) 2013-2017 Tatsuhiko Miyagawa.
# Copyright (c) 2017-2018 Marcel Greter.

use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;

################################################################################
# Emit a message to inform user of suboptimal watch behavior
# Filesys::Notify::KQueue will use inefficient polling scans
################################################################################

# stores choosen matcher
my $watchdeps = undef;

# list watchers per OS
my %watchers = (
	'linux' => [ 'Linux::Inotify2', 0.01 ],
	'cygwin' => [ 'Win32::ChangeNotify', 0.01 ],
	'mswin32' => [ 'Win32::ChangeNotify', 0.01 ],
	'darwin' => [ 'Filesys::Notify::KQueue', 0.01 ],
	'netbsd' => [ 'Filesys::Notify::KQueue', 0.01 ],
	'freebsd' => [ 'Filesys::Notify::KQueue', 0.01 ],
	'openbsd' => [ 'Filesys::Notify::KQueue', 0.01 ]
);

my $native_watcher = grep { /--native-watcher/ } @ARGV;
# check if dependency is wanted and/or available
if (exists $watchers{lc($^O)} && ! $ENV{PERL_FNS_NO_OPT}) {
	if ($native_watcher) { $watchdeps = $watchers{lc($^O)}; }
	elsif (!eval "require $watchers{lc($^O)}->[0]; return 1;") {
		warn "Consider installing $watchers{lc($^O)}->[0]\n";
		warn "Or configure with `--native-watcher` option\n";
	}
}

################################################################################
# See lib/ExtUtils/MakeMaker.pm for details of how to
# influence content of the Makefile that is written.
################################################################################

my %WriteMakefileArgs = (
	NAME               => "Filesys::Notify::Light",
	# finds $VERSION, requires EU::MM from perl >= 5.5
	VERSION_FROM       => 'lib/Filesys/Notify/Light.pm',
	# test dependencies
	TEST_REQUIRES      => {
		"File::Temp"       => 0,
		"Test::More"       => 0,
		"Test::SharedFork" => 0
	},
	# build dependencies
	BUILD_REQUIRES     => {
		'ExtUtils::MakeMaker'  => 6.30,
	},
	# build dependencies
	CONFIGURE_REQUIRES => {
		'ExtUtils::MakeMaker'  => 6.30,
	},
	# runtime dependencies
	PREREQ_PM => {
		# 'Getopt::Long'         => 0.01,
		# it doesn't have any dependencies
		# but you may want to install one of
		# Linux::Inotify2, Win32::ChangeNotify
		# Mac::FSEvents, Filesys::Notify::KQueue
		$watchdeps ? @{$watchdeps} : (),
	},
	# additional information
	META_MERGE => {
		resources => {
			license    => 'http://opensource.org/licenses/MIT',
			homepage   => 'https://metacpan.org/release/Filesys-Notify-Light',
			bugtracker => 'https://github.com/mgreter/Filesys-Notify-Light/issues',
			repository => 'https://github.com/mgreter/Filesys-Notify-Light',
		},
	},
	ABSTRACT           => "Light and simple file system watcher",
	# Original author, but now mostly rewritten completely
	AUTHOR             => q{Tatsuhiko Miyagawa <miyagawa@bulknews.net>},
	AUTHOR             => q{Marcel Greter <perl-file-notify@ocbnet.ch>},
	LICENSE            => "perl",
	EXE_FILES          => [],
);

# remove unknown key (as seen in Dist::Zilla)
unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
	my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
	my $br = $WriteMakefileArgs{BUILD_REQUIRES};
	for my $mod ( keys %$tr ) {
		if ( exists $br->{$mod} ) {
			if ($tr->{$mod} > $br->{$mod}) {
				$br->{$mod} = $tr->{$mod};
			}
		}
		else {
			$br->{$mod} = $tr->{$mod};
		}
	}
}

# remove unknown key (as seen in Dist::Zilla)
unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
	my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
	my $pp = $WriteMakefileArgs{PREREQ_PM};
	for my $mod ( keys %$br ) {
		if ( exists $pp->{$mod} ) {
			if ($br->{$mod} > $pp->{$mod}) {
				$pp->{$mod} = $br->{$mod};
			}
		}
		else {
			$pp->{$mod} = $br->{$mod};
		}
	}
}

# remove unknown key (as seen in Dist::Zilla)
unless ( eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
	delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
}

# See lib/ExtUtils/MakeMaker.pm for details of how to
# influence content of the Makefile that is written.
WriteMakefile(%WriteMakefileArgs);





lib/Filesys/Notify/Light.pm
package Filesys::Notify::Light;

use strict;
use warnings;

use 5.008_001;
our $VERSION = '0.20';
our $interval = 2.0;

# list watchers per OS
our %watchers = (
	'linux' => [ 'Linux::Inotify2', \&wait_inotify2, 0.01 ],
	'cygwin' => [ 'Win32::ChangeNotify', \&wait_win32, 0.01 ],
	'mswin32' => [ 'Win32::ChangeNotify', \&wait_win32, 0.01 ],
	'darwin' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
	'netbsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
	'freebsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
	'openbsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ]
);

# on cygwin we convert win to posix paths
use constant IS_CYGWIN => $^O eq "cygwin";

# core modules
use Cwd qw();
use Carp qw();
use File::Find qw();
use File::Basename qw();
use Fcntl qw(S_ISDIR);

# create new instance
sub new
{
	my($class, $path) = @_;
	Carp::croak('Usage: Filesys::Notify::Light->new([ $path1, $path2 ])')
		unless (ref $path eq 'ARRAY' && scalar(@_) == 2);
	my $self = bless { paths => $path }, $class;
	return $self->init;
}
# EO new

# choose file watcher
sub init
{
	my ($self) = @_;
	local $@; # preserver errors
	$self->{watcher_id} = "fullscan";
	$self->{watcher_cb} = \&wait_timer;
	return $self if $ENV{PERL_FNS_NO_OPT};
	if (my $watcher = $watchers{lc $^O}) {
		if (eval "require $watcher->[0]; 1") {
			$self->{watcher_id} = $watcher->[0];
			$self->{watcher_cb} = $watcher->[1];
		}
	}
	return $self;
}
# EO init

# main blocking wait method
sub wait
{
	my($self, $cb, $timeout) = @_;
	# store blocking timeout
	$self->{timeout} = $timeout;
	# instantiate the platform specific watcher function once
	$self->{watcher} ||= $self->{watcher_cb}->($self);
	# call watcher function
	$self->{watcher}->($cb);
}
# EO wait

# return observers struct
sub get_observers
{
	my @path = @_;

	my (%observer, @dirs, @files, @roots);
	@path = map { Cwd::abs_path($_) } @path;

	# split up watcher paths
	foreach my $path (@path) {
		if (-f $path) { push @files, $path; }
		elsif (-d $path) { push @dirs, $path; }
	}

	# get rid of inner paths
	# no need to observe twice
	foreach my $path (@dirs) {
		my $hasroot = undef;
		foreach my $root (@roots) {
			# check if path starts with root (is subpath)
			if (substr($path, 0, length($root)) eq $root) {
				$hasroot = $root;
				last;
			}
		}
		push @roots, $path unless $hasroot;
	}

	# observe directories recursively
	foreach my $path (@dirs) {
		if (exists $observer{$path}) {
			$observer{$path}->{isdir} = 1;
		} else {
			$observer{$path} = {
				files => [],
				isdir => 1
			};
		}
	}

	# observe files explicitly, but only if
	# base directory is not watched recursively
	foreach my $file (@files) {
		my $path = File::Basename::dirname($file);
		if (exists $observer{$path}) {
			push @{$observer{$path}->{files}}, $file;
		} else {
			$observer{$path} = {
				files => [$file],
				isdir => 0
			};
		}
	}

	# optimize scan targets for observer
	foreach my $path (keys %observer) {
		$observer{$path}->{scan} = $observer{$path}->{isdir}
			? [ $path ] : \ @{$observer{$path}->{files}};
	}

	return \ %observer;
}
# EO get_observers

# NetBSD, FreeBSD, OpenBSD and Mac OSX
sub wait_kqueue {

	my ($self) = @_;

	my %opt = (
		path => $self->{paths},
	);

	if (defined $self->{timeout}) {
		$opt{timeout} = $self->{timeout}
	}

	my $kqueue = Filesys::Notify::KQueue->new(%opt);

	return sub {
		my ($cb) = @_;
		$kqueue->wait($cb)
	};

}
# EO wait_kqueue

# Windows and CYGWIN
sub wait_win32 {

	my ($self) = @_;

	my $observers = get_observers(@{$self->{paths}});

	# either scan the whole directory or only the necessary files
	my @scan = map { @{$observers->{$_}->{scan}} } keys %{$observers};

	# get current filesystem state
	my $fs = _full_scan(@scan);

	my (@notify, @observer, @fs_state);

	for my $path (keys %{$observers}) {
		my $winpath = IS_CYGWIN ? Cygwin::posix_to_win_path($path) : $path;
		# 0x1b means 'DIR_NAME|FILE_NAME|LAST_WRITE|SIZE' = 2|1|0x10|8
		push @notify, Win32::ChangeNotify->new($winpath, $observers->{$path}->{isdir}, 0x1b);
		push @fs_state, _full_scan(@{$observers->{$path}->{scan}});
		push @observer, $observers->{$path};
	}

	return sub {

		my ($cb) = @_;

		my @events;
		while(1) {
			my $idx = Win32::ChangeNotify::wait_any(\@notify, $self->{timeout}); 
			Carp::croak("Can't wait notifications, maybe " . scalar(@notify) .
				" directories exceeds limitation.") if ! defined $idx;
			if($idx > 0) {
				--$idx;
				# get all file changes for observed path
				my $old_fs = $fs_state[$idx];
				my $observer = $observer[$idx];
				my $new_fs = _full_scan(@{$observer->{scan}});
				# on windows we can only watch folders
				# therefore we need to filter unwanted
				# events for files we are not looking for
				# but only if we don't watch folder itself
				unless ($observer->{isdir}) {
					foreach my $root (keys %{$new_fs}) {
						# process all reported file changes in path
						foreach my $file (keys %{$new_fs->{$root}}) {
							# don't remove if we watch particular file
							unless (exists $fs->{$root}->{$file}) {
								# we are not interested in this event
								delete $new_fs->{$root}->{$file};
							}
						}
					}
				}
				my $cb = sub { push @events, {
					path => $_[0],
					event => $_[1]
				}; };
				_compare_fs($old_fs, $new_fs, $cb);
				$fs_state[$idx] = $new_fs; # update state
				$notify[$idx]->reset; # reset watcher
				last if @events; # Actually changed
			}
		}
		$cb->(@events) if @events;
	}

}
# EO mk_wait_win32

# translate event types
sub _inotify2_type
{
	my ($event) = @_;
	return "create" if $event->IN_CREATE;
	return "delete" if $event->IN_DELETE;
	return "delete" if $event->IN_DELETE_SELF;
	return "modify"; # default case
}
# EO _inotify2_type

# Linux
sub wait_inotify2
{
	my ($self) = @_;

	my @path = @{$self->{paths}};

	my $wmode = 
		Linux::Inotify2::IN_MOVE() |
		Linux::Inotify2::IN_MODIFY() |
		Linux::Inotify2::IN_CREATE() |
		Linux::Inotify2::IN_DELETE() |
		Linux::Inotify2::IN_MOVE_SELF() |
		Linux::Inotify2::IN_DELETE_SELF();

	my $inotify = Linux::Inotify2->new;
	my $observers = get_observers(@path);

	my %watched;

	foreach my $observer (keys %{$observers}) {

		my @paths;

		if ($observers->{$observer}->{isdir}) {
			@paths = @{$observers->{$observer}->{scan}};
			@paths = keys %{_full_scan(@paths)};
		}
		else {
			@paths = @{$observers->{$observer}->{files}};
		}

		foreach my $path (@paths) {
			$inotify->watch($path, $wmode) or
				Carp::croak("watch failed: $!");
			$watched{$path} = 1;
		}

	}

	# return watcher sub
	return sub {
		my ($cb) = @_;
		$inotify->blocking(1);
		my @events = $inotify->read;
		foreach my $path (map { $_->fullname } @events) {
			next if exists $watched{$path} || ! -d $path;
			$inotify->watch($path, $wmode) or
				Carp::croak("watch failed: $!");
			$watched{$path} = 1;
		}
		$cb->(map { +{
			path => $_->fullname,
			event => _inotify2_type($_)
		} } @events);
	};

}
# EO wait_inotify2

# Pure perl fallback
sub wait_timer
{
	my ($self) = @_;

	my @path = @{$self->{paths}};

	my $fs = _full_scan(@path);

	# return watcher sub
	return sub {
		my ($cb) = @_;
		my @events;
		while (1) {
			# sleep 0 is needed to fix sigalrm on windows!?
			select(undef, undef, undef, $interval); sleep 0;
			my $new_fs = _full_scan(@path);
			my $cb = sub { push @events, {
				path => $_[0],
				event => $_[1]
			}; };
			_compare_fs($fs, $new_fs, $cb);
			$fs = $new_fs;
			last if @events;
		};
		$cb->(@events);
	};
}
# EO wait_timer

# compare fs states from old and new scan
# execute callback on all differences
sub _compare_fs
{
	my($old, $new, $cb) = @_;

	for my $dir (keys %$old) {
		for my $path (keys %{$old->{$dir}}) {
			if (!exists $new->{$dir}{$path}) {
				$cb->($path, 'delete');
			} elsif (!$new->{$dir}{$path}{is_dir} &&
					( $old->{$dir}{$path}{mtime} != $new->{$dir}{$path}{mtime} ||
					  $old->{$dir}{$path}{ctime} != $new->{$dir}{$path}{ctime} ||
					  $old->{$dir}{$path}{size}  != $new->{$dir}{$path}{size})) {
				$cb->($path, 'modify');
			}
		}
	}

	for my $dir (keys %$new) {
		for my $path (sort grep { !exists $old->{$dir}{$_} } keys %{$new->{$dir}}) {
			$cb->($path, 'create');
		}
	}
}
# EO _compare_fs

# scan all paths and return stats
sub _full_scan {
	my @paths = @_;

	my %map;
	for my $path (@paths) {
		next unless -e $path;
		# there is a weird race-condition on windows
		# checking hradlinks via lstat and stat seems to solve
		# the issue (File::find also does this internally)
		# only seems to happen with Win32::ChangeNotify
		# not sure if it is purely a timing issue
		my @stat = stat $path;
		my @lstat = lstat $path;
		next unless $stat[3];
		next unless $lstat[3];
		File::Find::finddepth({
			wanted => sub {
				my $fullname = $File::Find::fullname ||
					Cwd::abs_path($File::Find::name);
				if (my $stat = _stat($fullname)) {
					# keep track of directories to emit events for them
					$map{Cwd::abs_path($path)}{$fullname} = $stat if $stat->{is_dir};
					$map{Cwd::abs_path($File::Find::dir)}{$fullname} = $stat;
				}
			},
			follow_fast => 1,
			follow_skip => 2,
			no_chdir => 1,
		}, $path);

		# NOTE: On MSWin32, realpath and rel2abs disagree with path separator.
		# delete $map{$fp}{File::Spec->rel2abs($fp)} if exists $map{$fp};
	}

	return \%map;
}
# EO _full_scan

# get object with file/dir stats
sub _stat {
	my ($path) = @_;
	return unless -e $path;
	my @stat = stat $path;
	my @lstat = lstat $path;
	die "stat1: ", @stat unless $stat[3];
	die "stat2: ", @lstat unless $lstat[3];
	return unless $stat[3];
	return unless $lstat[3];
	return {
		path => $path,
		size => $stat[7],
		mtime => $stat[9],
		ctime => $stat[10],
		is_dir => S_ISDIR($stat[2]),
	};
}
# EO _stat

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Filesys::Notify::Light - Light and simple file system watcher

=head1 SYNOPSIS

  use Filesys::Notify::Light;

  my $watcher = Filesys::Notify::Light->new([ "." ]);
  $watcher->wait(sub {
	  for my $event (@_) {
		  $event->{path}; # full path of the file updated
		  $event->{event}; # type of the emitted event
	  }
  });

=head1 DESCRIPTION

Filesys::Notify::Light is a simple but unified interface to get
notifications of changes to a given filesystem path or for specific
files. It utilizes inotify2 on Linux, Win32::ChangeNotify on Windows,
KQueue on FreeBSD, OpenBSD, NetBSD and Mac OSX if they're installed,
with a fallback to a full directory scan if they're not available.

Implementation was forked from L<Filesys::Notify::Simple> in 2017 and
nearly fully rewritten in 2018 to fix most known issues and to improve
event emitting to include the type of the event (create, delete or modify).

=over 4

=item *

Currently C<wait> method blocks.

=back

In return, this module doesn't depend on any non-core modules.
Platform specific optimizations with L<Linux::Inotify2>,
L<Win32::ChangeNotify> or L<Filesys::Notify::KQueue> are optional.

NOTE: Using L<Win32::ChangeNotify> may put additional limitations.

=over 4

=item *

L<Win32::ChangeNotify> uses FindFirstChangeNotificationA (the A stands
for ANSI) so Unicode characters are not supported. On cygwin (1.7 or
later), Unicode characters should be able to be handled when
L<Win32::ChangeNotify> is not used.

=item *

If more than 64 directories are included under the specified paths,
an error occurs due to a limitation with the Windows API.

=item *

Directories are always watched recursively, since most observer
always report changes for the full directory tree. Filter out
events from subdirs if you don't want this behavior.

=back

=head1 AUTHOR

Marcel Greter E<lt>perl-file-notify@ocbnet.chE<gt>
Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<File::ChangeNotify> L<Filesys::Notify::Simple> L<Mac::FSEvents>
L<Linux::Inotify2> L<Filesys::Notify::KQueue> L<Win32::ChangeNotify>

=cutre

lib/Filesys/Notify/Simple.pm
package Filesys::Notify::Simple;

use strict;
use warnings;

use 5.008_001;
our $VERSION = '0.12';
our $interval = 2.0;

# list watchers per OS
our %watchers = (
    'linux' => [ 'Linux::Inotify2', \&wait_inotify2, 0.01 ],
    'cygwin' => [ 'Win32::ChangeNotify', \&wait_win32, 0.01 ],
    'mswin32' => [ 'Win32::ChangeNotify', \&wait_win32, 0.01 ],
    'darwin' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
    'netbsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
    'freebsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ],
    'openbsd' => [ 'Filesys::Notify::KQueue', \&wait_kqueue, 0.01 ]
);

use constant IS_CYGWIN => $^O eq "cygwin";

# core modules
use Cwd qw();
use Carp qw();
use File::Find qw();
use File::Basename qw();

# create new instance
sub new
{
    my($class, $path) = @_;
    Carp::croak('Usage: Filesys::Notify::Simple->new([ $path1, $path2 ])')
        unless (ref $path eq 'ARRAY' && scalar(@_) == 2);
    my $self = bless { paths => $path }, $class;
    return $self->init;
}
# EO new

# choose file watcher
sub init
{
    my ($self) = @_;
    local $@; # preserver errors
    $self->{watcher_cb} = \&wait_timer;
    return $self if $ENV{PERL_FNS_NO_OPT};
    if (my $watcher = $watchers{lc $^O}) {
        if (eval "require $watcher->[0]; 1") {
            $self->{watcher_cb} = $watcher->[1];
        }
    }
    return $self;
}
# EO init

# main blocking wait method
sub wait
{
    my($self, $cb, $timeout) = @_;
    # store blocking timeout
    $self->{timeout} = $timeout;
    # instantiate the platform specific watcher function once
    $self->{watcher} ||= $self->{watcher_cb}->($self);
    # call watcher function
    $self->{watcher}->($cb);
}
# EO wait

# return observers struct
sub get_observers
{
    my @path = @_;

    my (%observer, @dirs, @files, @roots);
    @path = map { Cwd::abs_path($_) } @path;

    # split up watcher paths
    foreach my $path (@path) {
        if (-f $path) { push @files, $path; }
        elsif (-d $path) { push @dirs, $path; }
    }

    # get rid of inner paths
    # no need to observer twice
    foreach my $path (@dirs) {
        my $hasroot = undef;
        foreach my $root (@roots) {
            # check if path starts with root (is subpath)
            if (substr($path, 0, length($root)) eq $root) {
                $hasroot = $root;
                last;
            }
        }
        push @roots, $path unless $hasroot;
    }

    # observe directories recursively
    foreach my $path (@dirs) {
        if (exists $observer{$path}) {
            $observer{$path}->{isdir} = 1;
        } else {
            $observer{$path} = {
                files => [],
                isdir => 1
            };
        }
    }

    # observe files explicitly, but only if
    # base directory is not watched recursively
    foreach my $file (@files) {
        my $path = File::Basename::dirname($file);
        if (exists $observer{$path}) {
            push @{$observer{$path}->{files}}, $file;
        } else {
            $observer{$path} = {
                files => [$file],
                isdir => 0
            };
        }
    }

    # optimize scan targets for observer
    foreach my $path (keys %observer) {
        $observer{$path}->{scan} = $observer{$path}->{isdir}
            ? [ $path ] : \ @{$observer{$path}->{files}};
    }

    return \ %observer;
}
# EO get_observers

# NetBSD, FreeBSD, OpenBSD and Mac OSX
sub wait_kqueue {
    my ($self) = @_;

    my %opt = (
        path => $self->{paths},
    );

    if (defined $self->{timeout}) {
        $opt{timeout} = $self->{timeout}
    }

    my $kqueue = Filesys::Notify::KQueue->new(%opt);

    return sub {
        my ($cb) = @_;
        $kqueue->wait($cb)
    };
}
# EO wait_kqueue

# Windows and CYGWIN
sub wait_win32 {

    my ($self) = @_;

    my $observers = get_observers(@{$self->{paths}});

    # either scan the whole directory or only the necessary files
    my @scan = map { @{$observers->{$_}->{scan}} } keys %{$observers};

    # get current filesystem state
    my $fs = _full_scan(@scan);

    my (@notify, @observer, @fs_state);

    for my $path (keys %{$observers}) {
        my $winpath = IS_CYGWIN ? Cygwin::posix_to_win_path($path) : $path;
        # 0x1b means 'DIR_NAME|FILE_NAME|LAST_WRITE|SIZE' = 2|1|0x10|8
        push @notify, Win32::ChangeNotify->new($winpath, $observers->{$path}->{isdir}, 0x1b);
        push @fs_state, _full_scan(@{$observers->{$path}->{scan}});
        push @observer, $observers->{$path};
    }

    return sub {

        my ($cb) = @_;

        my @events;
        while(1) {
            my $idx = Win32::ChangeNotify::wait_any(\@notify, $self->{timeout}); 
            Carp::croak("Can't wait notifications, maybe " . scalar(@notify) . " directories exceeds limitation.") if ! defined $idx;
            if($idx > 0) {
                --$idx;
                # get all file changes for observed path
                my $old_fs = $fs_state[$idx];
                my $observer = $observer[$idx];
                my $new_fs = _full_scan(@{$observer->{scan}});
                # on windows we can only watch folders
                # therefore we need to filter unwanted
                # events for files we are not looking for
                # but only if we don't watch folder itself
                unless ($observer->{isdir}) {
                    foreach my $root (keys %{$new_fs}) {
                        # process all reported file changes in path
                        foreach my $file (keys %{$new_fs->{$root}}) {
                            # don't remove if we watch particular file
                            unless (exists $fs->{$root}->{$file}) {
                                # we are not interested in this event
                                delete $new_fs->{$root}->{$file};
                            }
                        }
                    }
                }
                my $cb = sub { push @events, { path => $_[0], event => $_[1] }; };
                _compare_fs($old_fs, $new_fs, $cb);
                $fs_state[$idx] = $new_fs; # update state
                $notify[$idx]->reset; # reset watcher
                last if @events; # Actually changed
            }
        }
        $cb->(@events) if @events;
    }

}
# EO mk_wait_win32

# translate event types
sub _inotify2_type
{
    my ($event) = @_;
    return "create" if $event->IN_CREATE;
    return "delete" if $event->IN_DELETE;
    return "modify"; # default case
}
# EO _inotify2_type

# Linux
sub wait_inotify2
{
    my @path = @_;

    my $wmode = 
        Linux::Inotify2::IN_MOVE() |
        Linux::Inotify2::IN_MODIFY() |
        Linux::Inotify2::IN_CREATE() |
        Linux::Inotify2::IN_DELETE() |
        Linux::Inotify2::IN_MOVE_SELF() |
        Linux::Inotify2::IN_DELETE_SELF();

    my $inotify = Linux::Inotify2->new;
    my $observers = get_observers(@path);

    my %watched;

    foreach my $observer (keys %{$observers}) {

        my @paths;

        if ($observers->{$observer}->{isdir}) {
            @paths = @{$observers->{$observer}->{scan}};
            @paths = keys %{_full_scan(@paths)};
        }
        else {
            @paths = @{$observers->{$observer}->{files}};
        }

        foreach my $path (@paths) {
            $inotify->watch($path, $wmode) or
                Carp::croak("watch failed: $!");
            $watched{$path} = 1;
        }

    }

    # return watcher sub
    return sub {
        my ($cb) = @_;
        $inotify->blocking(1);
        my @events = $inotify->read;
        foreach my $path (map { $_->fullname } @events) {
            next if exists $watched{$path} || ! -d $path;
            $inotify->watch($path, $wmode) or
                Carp::croak("watch failed: $!");
            $watched{$path} = 1;
        }
        $cb->(map { +{ path => $_->fullname, event => _inotify2_type($_) } } @events);
    };

}
# EO wait_inotify2

# Pure perl fallback
sub wait_timer
{
    my ($self) = @_;

    my @path = @{$self->{paths}};

    my $fs = _full_scan(@path);

    # return watcher sub
    return sub {
        my ($cb) = @_;
        my @events;
        while (1) {
            # sleep 0 is needed to fix sigalrm on windows!?
            select(undef, undef, undef, $interval) && sleep 0;
            my $new_fs = _full_scan(@path);
            my $cb = sub { push @events, { path => $_[0], event => $_[1] }; };
            _compare_fs($fs, $new_fs, $cb);
            $fs = $new_fs;
            last if @events;
        };
        $cb->(@events);
    };
}
# EO wait_timer

# compare fs states
sub _compare_fs
{
    my($old, $new, $cb) = @_;

    for my $dir (keys %$old) {
        for my $path (keys %{$old->{$dir}}) {
            if (!exists $new->{$dir}{$path}) {
                $cb->($path, 'delete');
            } elsif (!$new->{$dir}{$path}{is_dir} &&
                    ( $old->{$dir}{$path}{mtime} != $new->{$dir}{$path}{mtime} ||
                      $old->{$dir}{$path}{size}  != $new->{$dir}{$path}{size})) {
                $cb->($path, 'modify');
            }
        }
    }

    for my $dir (keys %$new) {
        for my $path (sort grep { !exists $old->{$dir}{$_} } keys %{$new->{$dir}}) {
            $cb->($path, 'create');
        }
    }
}
# _compare_fs

sub _full_scan {
    my @paths = @_;

    my %map;
    for my $path (@paths) {
        my $fp = eval { Cwd::realpath($path) } or next;
        File::Find::finddepth({
            wanted => sub {
                my $fullname = $File::Find::fullname || File::Spec->rel2abs($File::Find::name);
                my $stat = $map{Cwd::realpath($File::Find::dir)}{$fullname} = _stat($fullname);
                $map{$path}{$fullname} = $stat if $stat && $stat->{is_dir}; # keep track of directories
            },
            follow_fast => 1,
            follow_skip => 2,
            no_chdir => 1,
        }, $path);

        # NOTE: On MSWin32, realpath and rel2abs disagree with path separator.
        # delete $map{$fp}{File::Spec->rel2abs($fp)} if exists $map{$fp};
    }

    return \%map;
}

sub _stat {
    my ($path) = @_;
    return undef unless -e $path;
    my @stat = stat $path;
    return {
        path => $path,
        mtime => $stat[9],
        size => $stat[7],
        is_dir => -d _,
    };
}


1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Filesys::Notify::Simple - Simple file system watcher

=head1 SYNOPSIS

  use Filesys::Notify::Simple;

  my $watcher = Filesys::Notify::Simple->new([ "." ]);
  $watcher->wait(sub {
      for my $event (@_) {
          $event->{path} # full path of the file updated
      }
  });

=head1 DESCRIPTION

Filesys::Notify::Simple is a simple but unified interface to get
notifications of changes to a given filesystem path or for specific
files. It utilizes inotify2 on Linux, Win32::ChangeNotify on Windows,
KQueue on FreeBSD, OpenBSD, NetBSD and Mac OSX if they're installed,
with a fallback to a full directory scan if they're not available.

There are some limitations in this module. If you don't like it, use
L<File::ChangeNotify>.

=over 4

=item *

Currently C<wait> method blocks.

=back

In return, this module doesn't depend on any non-core modules.
Platform specific optimizations with L<Linux::Inotify2>,
L<Win32::ChangeNotify> or L<Filesys::Notify::KQueue> are optional.

NOTE: Using L<Win32::ChangeNotify> may put additional limitations.

=over 4

=item *

L<Win32::ChangeNotify> uses FindFirstChangeNotificationA (the A stands
for ANSI) so Unicode characters are not supported. On cygwin (1.7 or
later), Unicode characters should be able to be handled when
L<Win32::ChangeNotify> is not used.

=item *

If more than 64 directories are included under the specified paths,
an error occurs due to a limitation with the Windows API.

=back

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<File::ChangeNotify> L<Mac::FSEvents> L<Linux::Inotify2> L<Filesys::Notify::KQueue>
L<Win32::ChangeNotify>

=cut

# directories are watched recursively